### PR TITLE
Issue-4882 Allow more than 16 collision shapes

### LINKED ISSUE #4882: Allow more than 16 shapes per collision object
**Is your feature request related to a problem? Please describe (REQUIRED):**
There is currently a hard limit of 16 shapes per collision object. This limit does not seem to be linked to any hard limit in Box2D or Bullet.

**Describe the solution you'd like (REQUIRED):**
The limit should either be configurable from game.project or the number of shapes of a collision object should be allowed to grow above the default 16 shapes.

**Describe alternatives you've considered (REQUIRED):**
Combine shapes into a convex hull shape.

**Additional context (OPTIONAL):**
https://forum.defold.com/t/support-for-more-than-16-shapes-in-a-collision-object/65471

### PR DESCRIPTION
**Pull request by britzl**

### CODE REVIEW

File: engine/gamesys/src/gamesys/resources/res_collision_object.cpp [hunk 1]
```diff
@@ -173,51 +177,61 @@ namespace dmGameSystem
...
-                            dmLogWarning("The collision object '%s' has a tile map containing more than %d layers, the rest will be ignored.", filename, COLLISION_OBJECT_MAX_SHAPES);
-                            shape_count = COLLISION_OBJECT_MAX_SHAPES;
-                        }
+                        uint32_t total_shapes_count = embedded_shape_count + shape_count;
+                        resource->m_Shapes2D.SetCapacity(total_shapes_count);
+                        resource->m_ShapeTranslation.SetCapacity(total_shapes_count);
+                        resource->m_ShapeRotation.SetCapacity(total_shapes_count);
+                        resource->m_Shapes2D.SetSize(total_shapes_count);
+                        resource->m_ShapeTranslation.SetSize(total_shapes_count);
+                        resource->m_ShapeRotation.SetSize(total_shapes_count);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_collision_object.cpp:189]: Thinking about this a bit more, there's really no need to have these as dynamic arrays.
We allocate them once, that's it.
I recommend replacing them with C arrays instead.

File: engine/gamesys/src/gamesys/resources/res_collision_object.cpp [hunk 2]
```diff
@@ -275,6 +282,16 @@ namespace dmGameSystem
                     dmPhysics::DeleteCollisionShape2D(resource->m_Shapes2D[i]);
                 }
             }
+            if (physics_context->m_3D)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_collision_object.cpp:285]: No need to check. It's valid to free a null pointer.
```
free(resource->m_Shapes2D);
free(resource->m_Shapes3D);
```

But never mind it here, the whole physics system is this mess of if-statements, checking 2d vs 3d :/

File: engine/gamesys/src/gamesys/resources/res_collision_object.cpp [hunk 3]
```diff
@@ -173,51 +177,53 @@ namespace dmGameSystem
...
+                resource->m_ShapeTranslation = (Vectormath::Aos::Vector3*)malloc(sizeof(Vectormath::Aos::Vector3) * embedded_shape_count);
+                resource->m_ShapeRotation = (Vectormath::Aos::Quat*)malloc(sizeof(Vectormath::Aos::Quat) * embedded_shape_count);
+            }
+            else if (!resource->m_TileGrid)
             {
-                dmLogWarning("Too many shapes in collision object. Up to %d is supported (%d). Discarding overflowing shapes.", COLLISION_OBJECT_MAX_SHAPES, count);
-                count = COLLISION_OBJECT_MAX_SHAPES;
+                resource->m_Shapes2D = (dmPhysics::HCollisionShape2D*)malloc(sizeof(dmPhysics::HCollisionShape2D) * embedded_shape_count);
+                resource->m_ShapeTranslation = (Vectormath::Aos::Vector3*)malloc(sizeof(Vectormath::Aos::Vector3) * embedded_shape_count);
+                resource->m_ShapeRotation = (Vectormath::Aos::Quat*)malloc(sizeof(Vectormath::Aos::Quat) * embedded_shape_count);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_collision_object.cpp:212]: This code is essentially in 3 places, when 1 should have sufficed.
But let's leave that refactoring until later.

**Review by JCash — APPROVED** [review]

File: engine/gamesys/src/gamesys/resources/res_collision_object.cpp (hunk 3 above)
Comment (britzl) [reply engine/gamesys/src/gamesys/resources/res_collision_object.cpp:212]: Yes, I'm not happy with how it looks.

**Pull request merged**